_posts/2024-04-12-optimizations.md

```diff
@@ -0,0 +1,771 @@
+the new counter value (`C + 1`) into `rcx`.
+
+```
+    mov rax, qword ptr [rbx+8]
+    mov rcx, qword ptr [rbx+16]
+```
+
+The next two instructions test whether the new counter value is a
+small integer that fits into a word. The test has been simplified to a
+more efficient test that is only safe when the value is known to be an
+integer. If it is a small integer, it is always safe to jump to the code
+that updates the existing tuple:
+
+```
+    test cl, 1
+    short je L38           ; Update directly if small integer.
+```
+
+If it is not a small integer, it must be a **bignum**, that is a
+signed integer that does not fit in 60 bits and therefore have to be
+stored on the heap with `rcx` containing a tagged pointer to the bignum on
+the heap.
+
+If `rcx` is a pointer to a term on the heap, it is not always safe to
+directly updating the existing tuple. That is because of the way the
+Erlang [generational garbage
+collector](https://www.erlang.org/doc/apps/erts/garbagecollection#generational-garbage-collection)
+works. Each Erlang process has two heaps for keeping Erlang terms:
+the young heap and the old heap. Terms on the young heap is allowed to
+reference terms on the old heap, but not vice versa. That means that
+if the record to be updated resides on the old heap, it is not allowed
+to store in it a reference to a term on the young heap.
+
+Therefore, the JIT needs to emit code to ensure that the point resides in the
+"safe part" of the young heap:
+
+```
+    mov rdi, [r13+480]     ; Get the high water mark
+    cmp rax, r15           ; Compare tuple pointer to heap top
+    short jae L39          ; Jump and copy if above
+    cmp rax, rdi           ; Compare tuple pointer to high water
+    short jae L38          ; Jump and overwrite if above high water
+```
+
+The safe part of the heap is between the high water mark and the heap
+top.  If the tuple is below the high water mark, if it is still alive,
+it will be copied to the old heap in the next garbage collection.
+
+If the tuple is in the safe part, the copy code is skipped by jumping
+to the code that stores the new value into the existing tuple.
+
+If not, the next part will copy the existing record to the heap.
+
+```
+L39:                       ; Copy the current record
+    vmovups ymm0, [rax-2]
+    vmovups [r15], ymm0
+    lea rax, qword ptr [r15+2] ; Set up tagged pointer to copy
+    add r15, 32            ; Advance heap top past the copy
+```
+
+The copying is done using [AVX instructions][avx].
+
+Next follows the code that writes the new value into the tuple:
+
+```
+L38:
+    mov rdi, rcx           ; Get new value for atoms field
+    mov qword ptr [rax+22], rdi
+    mov qword ptr [rbx+8], rax
+```
+
+[avx]: https://en.wikipedia.org/wiki/Advanced_Vector_Extensions
+
+If all the new values being written into the existing record are known
+never to be tagged pointers, the native instructions can be
+simplified. Consider this module:
```

Comment:
Should we mention that it's safe for literals too?

Reply:
Yes, good idea. I will make a new example that updates a record with a literal fun.